src/main/scala/edu/arizona/sista/odin/impl/TokenPatternCompiler.scala

```diff
@@ -133,6 +107,57 @@ class TokenPatternParsers(val unit: String) extends TokenConstraintParsers {
       case frag ~ n => frag.repeatPattern(n)
     }
 
+  // reverse grammar
+
+  def splitPatternRev: Parser[ProgramFragment] =
+    rep1sep(concatPatternRev, "|") ^^ { chunks =>
+      (chunks.head /: chunks.tail) {
+        case (lhs, rhs) =>
+          val split = Split(lhs.in, rhs.in)
+          ProgramFragment(split, lhs.out ++ rhs.out)
+      }
+    }
+
+  def concatPatternRev: Parser[ProgramFragment] =
+    rep1(quantifiedPatternRev) ^^ { chunks =>
+      (chunks.head /: chunks.tail) {
+        case (lhs, rhs) => ProgramFragment(rhs, lhs)
```

Comment:
Patterns are concatenated in reverse.
All other reverse productions are only needed because they call this one, either directly or indirectly